asynq/scheduler.py

```diff
@@ -60,262 +55,94 @@ def __init__(self):
 
     def reset(self):
-        """Resets the scheduler to its initial state."""
-        # Used to reverse the sequence of newly scheduled tasks
-        # to execute them in correct order
-        self._new_tasks = []
-        self._tasks = collections.deque()
         self._batches = set()
-        self.active_task = None
-
-    def start(self, task, run_immediately=False):
-        """Starts the specified task by simply scheduling it.
-        Returns the task passed to this method.
+        self._tasks = []
 
+    def await(self, task):
         """
-        if task.is_computed():
-            return task
-        if task.scheduler is not None:
-            raise AsyncTaskError('Task is already scheduled.')
-        if task.is_blocked():
-            raise AsyncTaskError("Task is blocked, so it shouldn't be scheduled.")
-
-        if run_immediately:
-            self._enqueue_new_tasks()
-            self._schedule(task)
-            self._continue_with_task()
-        else:
-            self._schedule(task)
-        return task
+        Executes a task and ensures it's complete when this method returns.
 
-    def continue_with(self, future, task):
-        """Starts (or continues) the task after specified future gets computed.
-        Returns the task passed to this method.
-
-        """
-        if task.is_computed():
-            raise AsyncTaskError('Task is already computed.')
-        if future.is_computed():
-            if task.scheduler is None:
-                return self.start(task)
-            else:
-                # Nothing to do in this case
-                return task
-        if task.scheduler is None:
-            task.scheduler = self
-        elif task.scheduler is not self:
-            raise AsyncTaskError('Task is already scheduled on another scheduler.')
-        self.add_dependency(task, future)
-        self._try_schedule(task, True)
-        return task
-
-    def flush_batches(self, must_await=False):
-        """Flushes all the batches that are currently 'known'
-        to scheduler. Normally you shouldn't call this method.
-
-        """
-        if must_await:
-            self.await()
-
-        error = None
-        while self._batches:
-            batches = self._batches
-            self._batches = set()
-            for batch in batches:
-                try:
-                    if batch.items and not batch.is_flushed():
-                        self._flush_batch(batch)
-                except BaseException as e:
-                    if error is None:
-                        error = e
-        if error is not None:
-            raise error
-        return 0
-
-    def await(self, tasks=[]):
-        """
-        Executes a list of tasks ensuring all of them are completed when this method returns.
-
-        :param tasks: tasks to await for
+        :param tasks: task to await for
         :return: ``None``
         """
-        self._await(tasks)
+        while not task.is_computed():
```

Comment:
If we call `await` on an async task and it doesn't end up yielding any batch items, we don't want to try flushing any batches (in case there's await recursion and we flush a batch from lower in the stack), so we check if it's computed after `_execute`. The check on line 74 shouldn't happen that often.